utils/sorting.hh

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <stdexcept>
+#include <vector>
+#include <map>
+#include <seastar/core/future.hh>
+#include "seastar/coroutine/maybe_yield.hh"
+#include "utils/stall_free.hh"
+
+namespace utils {
+
+/**
+ * Topological sort a DAG using Kahn's algorithm.
+ *
+ * @param vertices      Contains all vertices of the graph
+ * @param adjacency_map Entry {k, v} means there is an k->v edge in the graph
+ * @return a vector of topologically sorted vertices
```

Comment:
Nit: it would be good to clarify which one of the following is true about the final order:

- If there is an edge a -> b then a is before b in the result vector,
- If there is an edge a <- b then a is before b in the result vector.

The tests you introduce in the next commit suggest that the first statement is true.